addrmgr/netaddress.go

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2021 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package addrmgr
+
+import (
+	"encoding/base32"
+	"net"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// ServiceFlag identifies services supported by a Decred peer.
+type ServiceFlag uint64
```

Comment:
Why is this being duplicated here?  The supported services by a peer are not something defined by the address manager, rather they are part of the wire protocol.

Reply:
This was done to break the cyclical dependency between wire and addrmgr.  Otherwise, addrmgr would need to import wire for `wire.ServiceFlag` and wire would need to import addrmgr for `addrmgr.NetAddress`

Reply:
Why would `wire` depend on the new type though?  I would expect `wire` to continue to have its own protocol structure for sending information across the wire that would be converted as needed when going from the address manager's internal representation to wire representation and vice versa.

Please correct me if I'm wrong, but it seems to me that the primary aim of this PR is to "Decouple address manager from `wire.NetAddress", per the title, but having `wire` use the address manager's new `NetAddress` type means it just reverses the coupling.  Specifically, it would mean that `wire` then is coupled to `addrmgr.NetAddress` instead of `addrmgr` being coupled to `wire.NetAddress`, but the coupling would still be present.

I would expect the code in server such as `func (sp *serverPeer) pushAddrMsg` to accept the new `addrmgr.NetAddress` type and convert them in the loop that filters them.  Similarly, I would expect the code in `func (sp *serverPeer) OnVersion` to convert the `wire.NetAddress` to an `addrmgr.NetAddress`, etc.   That would ensure they are actually decoupled.

Reply:
Following up, I should also mention that wire *can't* change the type of the `addr` message without also bumping the major version, which we really, really, don't want to do because it would snowball into having to bump the major of every single API that uses **any** `wire` type, such as `wire.MsgBlock` and `wire.MsgTx` which is basically _everything_.  The fallout from that would be monumental.

Another thing is that the old `addr` message in `wire` will still need to be supported for a time, and it seems really odd to me that it would be defined in terms of something that supports things that the message itself doesn't.

Reply:
> Why would `wire` depend on the new type though? I would expect `wire` to continue to have its own protocol structure for sending information across the wire that would be converted as needed when going from the address manager's internal representation to wire representation and vice versa.

My intention here was to have `addrv2` depend on `addrmgr.NetAddress` so that the `Type` field could be included for each address in the message.  With the provided feedback in mind, it does seem like a new address type in `wire` could be introduce that is strictly used for serialization/deserialization for `addrv2`.  This way `addrmgr` still depends on wire, but only for `wire.ServiceFlag` and `wire` remains non-dependent on `addrmgr`.  Either way `addrmgr.ServiceFlag` can be removed in this PR as-is.

> I would expect the code in server such as `func (sp *serverPeer) pushAddrMsg` to accept the new `addrmgr.NetAddress` type and convert them in the loop that filters them. Similarly, I would expect the code in `func (sp *serverPeer) OnVersion` to convert the `wire.NetAddress` to an `addrmgr.NetAddress`, etc. That would ensure they are actually decoupled.

Yes, this is the way it is implemented now though the intention was to have `addrv2` depend on `addrmgr.NetAddress` rather than adding a new `wire` type for `addrv2`.  Additionally, my intent was to leave the `addr` implementation alone and dependent on `wire.NetAddress` with the mapping as described.

> Following up, I should also mention that wire can't change the type of the addr message without also bumping the major module version of wire, which we really, really, don't want to do because it would snowball into having to bump the major of every single API that uses any wire type, such as wire.MsgBlock and wire.MsgTx which is basically everything. The fallout from that would be monumental.

Right, the `wire` module has not been modified in this PR and the intent was to leave `addr` dependent on `wire.NetAddress`.  My takeaway here is that adding a new type would avoid the need to bump the `wire` module version for modules that do NOT consume the new type, is this a correct assumption to make?

Reply:
The addrmgr module has been modified to re-couple it to the wire module while also removing the dependency on wire.NetAddress, rather than inverting the dependency.  The mapping is still handled in server.go as it was before to convert the types across modules.